### ISSUE TITLE
[Head] Add Diffusion Policy action head

### ISSUE THREAD

**Issue by elprofesoriqo** [elprofesoriqo/FlowEdge#9]
## Goal

Add **Diffusion Policy** inference support to FlowEdge using a `ConditionalUnet1D` denoiser and fixed-memory DDIM/DDPM sampling.

The initial target is compatibility with the LeRobot `diffusion_pusht` policy while preserving FlowEdge's deterministic, allocation-free inference path.

## Why

Diffusion Policy is a widely used robotics policy architecture and gives FlowEdge a real external robotics workload beyond the current flow-matching smoke setup.

This implementation should also introduce reusable primitives for future diffusion- and sequence-based action heads.

## Initial support

The first implementation should support:

* `ConditionalUnet1D`
* 1D convolutional U-Net over the action horizon
* FiLM conditioning using observation condition + timestep embedding
* GroupNorm
* Mish activation
* DDIM sampling as the primary low-NFE path
* DDPM sampling for reference/parity
* fixed action horizon
* action normalization metadata
* action un-normalization after sampling
* C/C++ and Python-accessible inference path

Reference policy:

* LeRobot `diffusion_pusht`

## Out of scope

The first implementation does **not** need to support:

* arbitrary Diffusion Policy architectures
* training or fine-tuning
* dynamic action horizons
* dynamic tensor shapes
* accelerator-specific kernels
* generic diffusion image/model inference
* every scheduler supported by upstream libraries

The goal is one well-defined robotics policy path first.

### 1. Kernel primitives

Add reusable primitives behind the existing kernel abstraction:

* [ ] `conv1d`
* [ ] strided `conv1d` for downsampling
* [ ] `conv_transpose1d` for upsampling
* [ ] `group_norm`
* [ ] `mish`
* [ ] FiLM conditioning
* [ ] timestep embedding

Start with a correct scalar implementation.

AVX2/NEON optimization should be added only where useful and benchmarked separately.

### 2. Diffusion denoiser

Add:

```text
heads/diffusion/
├── diffusion.h
└── diffusion.cc
```

Implement the fixed-shape `ConditionalUnet1D` inference path.

Requirements:

* [ ] all runtime buffers are known at initialization
* [ ] no heap allocation inside denoiser execution
* [ ] fixed tensor layouts are documented
* [ ] intermediate shapes are validated during model loading

### 3. DDIM / DDPM sampler

Implement:

* [ ] precomputed diffusion schedule
* [ ] DDIM sampling
* [ ] DDPM sampling
* [ ] deterministic fixed-seed execution
* [ ] O(1)-allocation sampling loop

Sampler state required during inference should be preallocated in the arena.

### 4. Model conversion

Extend `convert/` with a documented mapping for:

```text
LeRobot diffusion_pusht
        ↓
FlowEdge Diffusion Policy
```

The converter should:

* [ ] map supported `ConditionalUnet1D` weights
* [ ] validate required tensors and dimensions
* [ ] reject unsupported variants with a useful error
* [ ] preserve action normalization metadata
* [ ] preserve action horizon / action dimension metadata

### 5. Action normalization

Carry the policy's normalization information through conversion.

The runtime should:

* [ ] consume normalized action-space samples internally
* [ ] apply the corresponding action un-normalization before returning actions
* [ ] support the normalization scheme used by the reference `diffusion_pusht` policy

### 6. Public API

Expose the new head through the existing runtime interfaces.

Required:

* [ ] C ABI sampling entry point
* [ ] action horizon reporting
* [ ] action dimension reporting
* [ ] Python binding
* [ ] model/head identification through checkpoint metadata

### 7. Reference implementation and verification

Add a minimal PyTorch reference implementation:

```text
torch_ref.py
```

Integrate it with the existing verification workflow.

Verify:

* [ ] denoiser output
* [ ] intermediate primitives where useful
* [ ] fixed-seed DDIM trajectory
* [ ] fixed-seed DDPM trajectory
* [ ] final action output

Parity tolerances should be explicitly documented rather than relying only on a generic pass/fail comparison.

## Tests

At minimum:

* [ ] Mish vs reference
* [ ] GroupNorm vs reference
* [ ] Conv1D vs naive/reference implementation
* [ ] transposed Conv1D vs reference
* [ ] FiLM correctness
* [ ] timestep embedding correctness
* [ ] denoiser parity
* [ ] deterministic sampling for fixed seed/input
* [ ] sampler output remains finite
* [ ] DDIM and DDPM follow distinct sampling paths
* [ ] normalization / un-normalization round trip
* [ ] model conversion validation
* [ ] end-to-end reference policy inference

## Example

Add a minimal example under:

```text
examples/diffusion_sample.*
```

The example should demonstrate:

```text
observation condition
        ↓
ConditionalUnet1D
        ↓
DDIM sampler
        ↓
un-normalization
        ↓
action horizon
```

Keep the example focused on inference rather than preprocessing or environment integration.

## Performance requirements

After initialization:

* no heap allocations during sampling;
* diffusion schedule remains precomputed;
* fixed-shape intermediate buffers are reused;
* performance-sensitive kernel changes should include benchmark results.

Benchmark at minimum:

* denoiser forward latency
* one complete DDIM action generation
* p50
* p95
* p99

Performance comparisons should include CPU, compiler, build type and number of sampling steps.

## Acceptance criteria

This issue is complete when:

* [ ] the documented LeRobot `diffusion_pusht` checkpoint can be converted
* [ ] the converted model loads successfully in FlowEdge
* [ ] fixed input + fixed initial noise produce output within documented tolerance of the PyTorch reference
* [ ] DDIM inference is deterministic for a fixed input and seed
* [ ] sampling performs no heap allocation after initialization
* [ ] action normalization metadata survives conversion
* [ ] returned actions are correctly un-normalized
* [ ] C/C++ API works
* [ ] Python API works
* [ ] unit and end-to-end tests pass
* [ ] a minimal example is included
* [ ] supported limitations are documented

## Suggested child issues

This epic should be split into independently reviewable tasks:

* [ ] **Kernel:** Add scalar Mish activation + tests
* [ ] **Kernel:** Add GroupNorm + tests
* [ ] **Kernel:** Add Conv1D / strided Conv1D primitives
* [ ] **Kernel:** Add ConvTranspose1D primitive
* [ ] **Kernel:** Add FiLM + timestep embedding
* [ ] **Head:** Implement `ConditionalUnet1D`
* [ ] **Sampler:** Implement fixed-memory DDIM/DDPM
* [ ] **Convert:** Add LeRobot Diffusion Policy converter
* [ ] **Verify:** Add PyTorch reference fixtures and parity tests
* [ ] **API:** Expose Diffusion Policy through C/Python
* [ ] **Example:** Add end-to-end Diffusion Policy sample

## References

* LeRobot `diffusion_pusht`
* https://huggingface.co/lerobot/diffusion_pusht

**Comment by elprofesoriqo** [comment]
Issue #9 is now split into a reviewable stack:

1. #28 - runtime and scalar diffusion kernels (`feat/issue-9-diffusion-core`)
2. #29 - LeRobot converter, C/Python APIs, examples, and parity checks (`feat/issue-9-diffusion-converter`)
3. #30 - focused latency/kernel benchmarks and install-consumer coverage (`perf/issue-9-diffusion-bench`)
4. #31 - action contracts, converter guide, API docs, and benchmark evidence (`docs/issue-9-diffusion`)

The branches are intentionally based on one another so each PR remains small. Please review/merge in order. The next existing work queue is #23 (checkpoint preflight), #24 (deployment profile), and #25 (deadline/regression benchmark profile).

**Comment by elprofesoriqo** [comment]
Diffusion Policy support is complete on main through the merged stack: #28 runtime/kernels, #29 converter + C/Python APIs + parity fixtures, #30 focused benchmarks/install coverage, #31 contracts/docs, and #59 Diffusion Policy checkpoint preflight. Verification includes deterministic DDIM/DDPM parity, action normalization, fixed-memory runtime behavior, examples, and cross-platform CI/Docker coverage. Closing this completed epic; future architecture variants should be tracked as new issues.
